packages/ses/src-xs/compartment.js

```diff
@@ -0,0 +1,284 @@
+/* eslint-disable @endo/no-nullish-coalescing */
+import {
+  Map,
+  Object,
+  SyntaxError,
+  TypeError,
+  WeakMap,
+  arrayMap,
+  create,
+  defineProperty,
+  entries,
+  fromEntries,
+  getOwnPropertyDescriptor,
+  globalThis,
+  mapDelete,
+  mapGet,
+  mapSet,
+  uncurryThis,
+  weakmapGet,
+  weakmapSet,
+} from '../src/commons.js';
+import { nativeGetGlobalThis } from './commons.js';
+import {
+  makeCompartmentConstructor,
+  compartmentOptions,
+} from '../src/compartment.js';
+import { getGlobalIntrinsics } from '../src/intrinsics.js';
+import { tameFunctionToString } from '../src/tame-function-tostring.js';
+import { chooseReporter } from '../src/reporting.js';
+
+const muteReporter = chooseReporter('none');
+
+export const ShimStartCompartment = makeCompartmentConstructor(
+  makeCompartmentConstructor,
+  getGlobalIntrinsics(globalThis, muteReporter),
+  tameFunctionToString(),
+);
+
+export const shimCompartmentPrototype = ShimStartCompartment.prototype;
+
+export const shimEvaluate = uncurryThis(shimCompartmentPrototype.evaluate);
+export const shimImport = uncurryThis(shimCompartmentPrototype.import);
+export const shimImportNow = uncurryThis(shimCompartmentPrototype.importNow);
+
+export const shimGetGlobalThis = uncurryThis(
+  getOwnPropertyDescriptor(ShimStartCompartment.prototype, 'globalThis').get,
+);
+
+export const privateFields = new WeakMap();
+
+const adaptVirtualModuleSource = ({
+  execute,
+  imports = [],
+  exports = [],
+  reexports = [],
+}) => {
+  const resolutions = create(null);
+  let i = 0;
+  return {
+    execute(environment) {
+      const fakeCompartment = {
+        importNow(specifier) {
+          return environment[specifier];
+        },
+      };
+      execute(environment, fakeCompartment, resolutions);
+    },
+    bindings: [
+      ...arrayMap(imports, specifier => {
+        const resolved = `_${i}`;
+        i += 1;
+        resolutions[specifier] = resolved;
+        return { importAllFrom: specifier, as: resolved };
+      }),
+      ...arrayMap(reexports, specifier => ({ exportAllFrom: specifier })),
+      ...arrayMap(exports, name => ({ export: name })),
+    ],
+  };
+};
+
+const adaptModuleSource = source => {
+  if (source.execute) {
+    return adaptVirtualModuleSource(source);
+  }
+  // eslint-disable-next-line no-underscore-dangle
+  if (source.__syncModuleProgram__) {
+    throw new SyntaxError(
```

Comment:
I'm pretty sure @erights prefers universal avoidance of `new` with built-in error constructors.